### PR TITLE
Fix/player 3832

### PR DESCRIPTION
**Pull request by zelibobla**
Implemented debouncing of ads requests to prevent Google IMA 1009 error

### CODE REVIEW

File: js/google_ima.js [hunk 1]
```diff
@@ -524,13 +526,33 @@ require("../html5-common/js/utils/utils.js");
...
+      let debouncerId = null;
       this.playAd = function(amcAdPod)
       {
+        if (debouncerId) {
+          _resetAdsState();
+        }
+        debouncerId = setTimeout(() => {
+          debouncerId = null;
+          playAd(amcAdPod);
+        }, DEBOUNCE_TIMEOUT);
```
Comment (aeng7) [inline js/google_ima.js]: Pasting my comment from the JIRA ticket:

Is this adding 500 ms to the user experience prior to an ad playback (even a single ad playback), or will this only be used if there are multiple requests in quick succession? If for a single preroll/ad there is now an additional 500 ms of latency, we cannot use this fix/workaround. That adds a significant time to time to first frame for prerolls and when transitioning from content to ads.

File: js/google_ima.js [hunk 2]
```diff
@@ -524,13 +526,33 @@ require("../html5-common/js/utils/utils.js");
       });
 
       /**
-       * Called by the ad manager controller.  Ad Manager Controller lets the module know that an ad should play now.
-       * @public
+       * debouncing the case when user forwarded ahead and 2 or more ads requests simultaneosly issued
```
Comment (pilievOoyala) [inline js/google_ima.js]: Debouncing the case when a user forwarded ahead and 2 or more ad requests are simultaneously issued

File: js/google_ima.js [hunk 3]
```diff
@@ -524,13 +526,33 @@ require("../html5-common/js/utils/utils.js");
...
+       * @see https://jira.corp.ooyala.com/browse/PLAYER-3832
```
Comment (pilievOoyala) [inline js/google_ima.js]: Lets keep tickets out of the jsdocs.  External people can't see our tickets. You can add it as a note above the code in the function.

File: js/google_ima.js [hunk 4]
```diff
@@ -524,13 +526,33 @@ require("../html5-common/js/utils/utils.js");
...
+        }, DEBOUNCE_TIMEOUT);
+      }
+
+      /**
+       * Called by the ad manager controller.  Ad Manager Controller lets the module know that an ad should play now.
+       * @private
+       * @method GoogleIMA#playAd
+       * @param {object} ad The ad to play from the timeline.
+       */
+      let playAd = (amcAdPod) => {
```
Comment (pilievOoyala) [inline js/google_ima.js]: please use the old style of function declaration

Comment (KseniiaSokolovaEpam) [reply js/google_ima.js]: @pilievOoyala oh, Peter! I am so pity to hear that we have to use old style! Why? 
ES6 is good practice is not it?
... And we use it already in the plugins, skin and mjolnir ...

Comment (aeng7) [reply js/google_ima.js]: I also agree we should be using ES6. @pilievOoyala what is the reason we should use the old style of function declaration?

Comment (pilievOoyala) [reply js/google_ima.js]: You guys have convinced me. I see the advantages of the new style and not needing to do a bind.

File: js/google_ima.js (hunk 1 above)
Comment (aeng7) [reply js/google_ima.js]: @zelibobla please take a look at this comment. This is a major issue if we are adding 500 ms prior to every ad playback.

Comment (zelibobla) [reply js/google_ima.js]: Yep, we had to have a call for this, but due to another priority this wasn't discussed. Just forgot then.
At the moment when user clicked forward and got over the ads (and we insist him to see the ads), we yet don't know if he will click immediately again and get over next ads. This is why we need debounce timeout. And yes: we have to use it even for single ads.
It hits UX yes, as I explained originally before implementing this.

File: js/google_ima.js (hunk 3 above)
Comment (zelibobla) [reply js/google_ima.js]: Removed

**Comment by zelibobla** [comment]
Closing as outdated

**Pull request closed without merging**